library/nist_kw.c

```diff
@@ -467,7 +475,7 @@ int mbedtls_nist_kw_unwrap( mbedtls_nist_kw_context *ctx,
 cleanup:
     if( ret != 0 )
     {
-        mbedtls_platform_zeroize( output, *out_len - 1 );
+        memset( output, 0, *out_len );
```

Comment:
That's still not right: you have `*out_len = 0` on many error paths even though some data was written to the buffer.

Reply:
Every error path that some data has been written to the buffer, updates `out_len` AFAIK. Whether it is in `mbedtls_cipher_update ()` (`out_len` is updated to 16 ) or through the `unwrap()` helper function. If `unwrap()` succeeds, but authentication fails, `out_len` will already be updated. This is why I keep setting `*out_len = 0` the line after.
If `unwrap()` fails, then the buffer will already be cleared within, and `out_len` set to 0.
Did you find a path where this is not so?

Reply:
I had a closer look at `unwrap` and it's worse than that: it writes past the maximum size of the output buffer, not just past `*out_len` at a given time.